src/sc/sidechain.cpp

```diff
@@ -632,3 +448,338 @@ void ScMgr::dump_info()
+
+        LogPrint("sc", "%s():%d - immature amount after: %s\n",
+            __func__, __LINE__, FormatMoney(iaMap[maturityHeight]));
+
+        if (iaMap[maturityHeight] < 0)
+        {
+            // should not happen either 
+            LogPrint("sc", "ERROR %s():%d - scId=%s negative balance at height=%d\n",
+                __func__, __LINE__, scId.ToString(), maturityHeight);
+            return false;
+        }
+
+        if (iaMap[maturityHeight] == 0)
+        {
+            iaMap.erase(maturityHeight);
+            LogPrint("sc", "%s():%d - removed entry height=%d from immature amounts in memory\n",
+                __func__, __LINE__, maturityHeight );
+        }
+        // mark this sc to be flushed in memory
+        sDirty.insert(scId);
+    }
+
+    // remove sidechain if the case
+    BOOST_FOREACH(const auto& entry, tx.vsc_ccout)
+    {
+        const uint256& scId = entry.scId;
+
+        LogPrint("sc", "%s():%d - removing scId=%s\n", __func__, __LINE__, scId.ToString());
+
+        const auto it_map = mUpdate.find(scId);
+        if (it_map == mUpdate.end() )
+        {
+            // should not happen 
+            LogPrint("sc", "ERROR: %s():%d - scId=%s not in scView\n", __func__, __LINE__, scId.ToString() );
+            return false;
+        }
+ 
+        if (it_map->second.balance > 0)
+        {
+            // should not happen either 
+            LogPrint("sc", "ERROR %s():%d - scId=%s balance not null: %s\n",
+                __func__, __LINE__, scId.ToString(), FormatMoney(it_map->second.balance));
+            return false;
+        }
+ 
+        if (mUpdate.erase(scId))
+        {
+            LogPrint("sc", "%s():%d - scId=%s removed from scView\n", __func__, __LINE__, scId.ToString() );
+        }
+
+        // do not update in memory
+        sDirty.erase(scId);
+
+        // mark this sc to be removed from in memory
+        sErase.insert(scId);
+    }
+    return true;
+}
+
+bool ScCoinsViewCache::ApplyMatureBalances(int blockHeight, CBlockUndo& blockundo)
+{
+    LogPrint("sc", "%s():%d - blockHeight=%d, msc_iaundo size=%d\n", __func__, __LINE__, blockHeight,  blockundo.msc_iaundo.size() );
+
+    auto it_map = mUpdate.begin();
+
+    while (it_map != mUpdate.end() )
+    {
+        const uint256& scId = it_map->first;
+        ScInfo& info = it_map->second;
+        const std::string& scIdString = scId.ToString();
+
+        auto it_ia_map = info.mImmatureAmounts.begin();
+
+        while (it_ia_map != info.mImmatureAmounts.end() )
+        {
+            int maturityHeight = it_ia_map->first;
+            CAmount a = it_ia_map->second;
+
+            LogPrint("sc", "%s():%d - adding immature balance entry into blockundo (h=%d, amount=%s) for scid=%s\n",
+                __func__, __LINE__, maturityHeight, FormatMoney(a), scIdString);
+
+            // store immature balances into the blockundo obj
+            blockundo.msc_iaundo[scId][maturityHeight] += a;
```

Comment:
why `+= a` and not ` = a`? Seems that for `blockundo` specific sidechain on specific maturity height amount should be set only once, because now we accumulate  `mImmatureAmounts` by maturity height.

Reply:
Ok, correct, will apply the modification